lib/css-converter.js

```diff
@@ -0,0 +1,257 @@
+  // The no-useless-escape regex rule is wrong when it's in a Regex. These escapes are intentional.
+  // eslint-disable-next-line no-useless-escape
+  return str.replace(/[\[\^\$\.\|\?\*\+\(\)\\]/g, (tok) => `\\${tok}`);
+}
+
+/**
+ * Add android:id/ to beginning of string if it's not there already
+ * @param {*} str
+ */
+function prependAndroidId (str) {
+  if (!str.startsWith('android:id/')) {
+    return `android:id/${str}`;
+  }
+  return str;
+}
+
+/**
+ * Convert a CSS attribute into a UiSelector method call
+ * @param {*} cssAttr CSS attribute object
+ */
+function parseAttr (cssAttr) {
+  if (cssAttr.valueType && cssAttr.valueType !== 'string') {
+    log.errorAndThrow(`Could not parse '${cssAttr.name}=${cssAttr.value}'.` +
+      `Unsupported attribute type '${cssAttr.valueType}'`);
+  }
+  const attrName = assertGetAttrName(cssAttr);
+  const methodName = toSnakeCase(attrName);
+  if (booleanAttrs.includes(attrName)) {
+    return `.${methodName}(${assertGetBool(cssAttr)})`;
+  }
+
+  if (strAttrs.includes(attrName)) {
+    let value = cssAttr.value || '';
+    if (attrName === 'resource-id') {
+      value = prependAndroidId(value);
+    }
+    if (value === '') {
+      return `.${methodName}Matches("")`;
+    } else {
+      switch (cssAttr.operator) {
+        case '=':
+          return `.${methodName}("${value}")`;
+        case '*=':
+          if (['description', 'text'].includes(attrName)) {
+            return `.${methodName}Contains("${value}")`;
+          }
+          return `.${methodName}Matches("${escapeRegexLiterals(value)}")`;
+        case '^=':
+          if (['description', 'text'].includes(attrName)) {
+            return `.${methodName}StartsWith("${value}")`;
+          }
+          return `.${methodName}Matches("^${escapeRegexLiterals(value)}")`;
+        case '$=':
+          return `.${methodName}Matches("${escapeRegexLiterals(value)}$")`;
+        case '~=':
+          return `.${methodName}Matches("${getWordMatcherRegex(value)}")`;
+        default:
+          log.errorAndThrow(`Unsupported CSS attribute operator '${cssAttr.operator}'`);
+      }
+    }
+  }
+}
+
+/**
+ * Convert a CSS pseudo class to a UiSelector
+ * @param {*} cssPseudo CSS Pseudo class
+ */
+function parsePseudo (cssPseudo) {
+  if (cssPseudo.valueType && cssPseudo.valueType !== 'string') {
+    log.errorAndThrow(`Could not parse '${cssPseudo.name}=${cssPseudo.value}'. ` +
+      `Unsupported css pseudo class value type: '${cssPseudo.valueType}'`);
+  }
+
+  const pseudoName = assertGetAttrName(cssPseudo);
+
+  if (booleanAttrs.includes(pseudoName)) {
+    return `.${toSnakeCase(pseudoName)}(${assertGetBool(cssPseudo)})`;
+  }
+
+  if (numericAttrs.includes(pseudoName)) {
+    return `.${pseudoName}(${cssPseudo.value})`;
+  }
+}
+
+/**
+ * Convert a CSS rule to a UiSelector
+ * @param {*} cssRule CSS rule definition
+ */
+function parseCssRule (cssRule) {
+  let uiAutomatorSelector = 'new UiSelector()';
```

Comment:
Did you think about having `xpath` instead of UiSelector? With xpath we could also easily migrate this functionality into xcuitest or Espresso. Although, UiSelector is only specific to UIAutomator

Reply:
@mykola-mokhnach I chose `UiSelector` because I want this to be a selector that gets native performance. I want to apply this same code to iOS and have it translate to `class chain` and to Espresso and have it translate to the ViewMatcher and DataMatcher selectors.